signer/src/util.rs

```diff
@@ -60,6 +60,76 @@ pub struct InnerApiFallbackClient<T> {
     retry_count: AtomicU8,
 }
 
+/// A context that provides information about the current retry attempt and
+/// allows the caller to abort the retry loop.
+#[derive(Debug, Clone)]
+pub struct RetryContext {
+    inner: Arc<InnerRetryContext>,
+}
+
+/// Inner implementation of the retry context.
+#[derive(Debug)]
+pub struct InnerRetryContext {
+    /// The total number of _retries_ that will be attempted. Note that this
+    /// is one less than the total number of attempts, as the first attempt
+    /// is implicit.
+    total_retries: u8,
+    /// The current _retry attempt_. Note that this is zero-indexed, so the
+    /// first retry attempt is `0` which represents the initial (implicit)
+    /// attempt.
+    current_retry: u8,
+    /// If set to true, the current retry loop will be aborted.
+    abort: AtomicBool,
+}
+
+impl RetryContext {
+    /// Create a new retry context.
+    ///
+    /// The `total_retries` parameter is the total number of **retries** that will
+    /// be attempted, _excluding_ the initial attempt.
+    ///
+    /// The `current_retry` parameter is the current retry attempt, starting at `0`
+    /// which represents the initial (implicit) attempt.
+    fn new(total_retries: u8, current_retry: u8) -> Self {
+        Self {
+            inner: Arc::new(InnerRetryContext {
+                total_retries,
+                current_retry,
+                abort: AtomicBool::new(false),
+            }),
+        }
+    }
+
+    /// Gets the total number of attempts that will be made.
+    pub fn total_attempts(&self) -> u8 {
+        self.inner.total_retries + 1
+    }
+
+    /// Gets the current attempt number.
+    pub fn current_attempt(&self) -> u8 {
+        self.inner.current_retry + 1
+    }
+
+    /// If a client call fails, this method can be used to abort the retry loop
+    /// and return the current result immediately.
+    pub fn abort(&self) {
+        self.inner.abort.store(true, Ordering::SeqCst);
+    }
```

Comment:
IIUC so far we only need the retry context/abort mechanism to return errors that are not retriable errors.

If so, an alternative approach to this could be to wrap return such errors in a `RetriableError` of some kind, and in `exec` check for such errors to continue the loop or return. Doing so, it would make it more explicit what can be retried and what is an actual error (and maybe reusable for errors not originating from the closure itself), vs now that we need to remember to call `abort`.

Reply:
Yeah that was an option which I had considered as well, but I don't think it would be as nice to use, you still have to remember to use a `RetriableError` (from your example) and in the majority of cases you pretty much always want to retry anyway (for `GET`-style requests, etc.).

But iirc the bigger issue was that unless we wanted to "infect" `Error` with generics to be able to use the pass-through nature of the result of the closure we would have to use `RetriableError(Box<dyn std::error::Error>)`, but then that kills the ability to match on what comes out of `exec()`, which to me is a pretty big downside.